transformer_engine/pytorch/cpu_offload.py

```diff
@@ -0,0 +1,449 @@
+    :func:`~torch.autograd.function._ContextMethodMixin.save_for_backward` but
+    also those recorded by a PyTorch-defined operation).
+
+    The ``on_get_saved_tensors`` method will be called when the backward function
+    of this op attempts to retrieve the saved tensor from context (this includes
+    :func: `torch.Tensor.backward()` or :func: `torch.autograd.grad()`. It takes the
+    as input the return value of the ``on_save_for_backward``, and is meant to return
+    an identical copy of the tensor being saved by ``on_save_for_backward`` in terms of
+    size, device and element values.
+
+    Example:
+
+        >>> import torch
+        >>> from typing import Any
+        >>>
+        >>> class DummyHook(CpuOffloadSavedTensorHook):
+        ...
+        ...     def on_save_for_backward(self, tensor: torch.Tensor) -> Any:
+        ...         logging.info("On save", tensor)
+        ...         return (tensor,)
+        ...
+        ...     def on_get_saved_tensor(self, saved_state: Any) -> torch.Tensor:
+        ...         logging.info("On get", saved_state)
+        ...         tensor, = saved_state
+        ...         return tensor
+        ...
+        >>> a = torch.ones(5, requires_grad=True)
+        >>> b = torch.ones(5, requires_grad=True) * 2
+        >>> with DummyHook():
+        ...     y = a * b
+        ...
+        On save tensor([1., 1., 1., 1., 1.], requires_grad=True)
+        On save tensor([2., 2., 2., 2., 2.], grad_fn=<MulBackward0>)
+        >>> y.sum().backward()
+        On get (tensor([1., 1., 1., 1., 1.], requires_grad=True),)
+        On get (tensor([2., 2., 2., 2., 2.], grad_fn=<MulBackward0>),)
+
+    """
+
+    def __init__(self) -> None:
+        self.inside_context = False
+
+    def __enter__(self):
+        global CPUOffloadEnabled
+        CPUOffloadEnabled = True
+
+        self.inside_context = True
+        torch._C._autograd._push_saved_tensors_default_hooks(
+            self.on_save_for_backward,
+            self.on_get_saved_tensor
+            )
+
+    def __exit__(self, *args: Any):
+        global CPUOffloadEnabled
+        CPUOffloadEnabled = False
+
+        self.inside_context = False
+        torch._C._autograd._pop_saved_tensors_default_hooks()
+
+
+    def on_save_for_backward(self, tensor: torch.Tensor) -> Any:
+        raise NotImplementedError("`on_save_for_backward: Callable[[torch.Tensor], Any]`"
+                                  "is not implemented in CpuOffloadHook class. Inherit "
+                                  "this class and implement your custom hooks")
+
+    def on_get_saved_tensor(self, saved_state: Any) -> torch.Tensor:
+        raise NotImplementedError("`on_get_saved_tensors: Callable[[Any], torch.Tensor]`"
+                                  "is not implemented in CpuOffloadHook class. Inherit "
+                                  "this class and implement your custom hooks")
+
+class CpuOffloadHookWithOffloadHandler(CpuOffloadSavedTensorHook):
```

Comment:
This seems to be a general functionality instead of it being specific to CPU offloading, could we rename it as such?

Reply:
Additionally, is this a reimplementation of `torch.autograd.graph.saved_tensors_hooks`?

Reply:
>>Additionally, is this a reimplementation of torch.autograd.graph.saved_tensors_hooks?

Yes, this is a modified version of the hook you mentioned but not a re-implementation. We wanted to maintain state across different calls to this hook, so we made it as a Class object.

Reply:
>This seems to be a general functionality instead of it being specific to CPU offloading, could we rename it as such?

I don't understand how do you call it generic, because this Class needs to get a offload handler based on which offloading happens. The offload handler handles asynchronously launching CUDA streams. For forward, it offloads activations/weights of the previously computed TF layer while for backward it reloads the next layers weight/activations. 

The offload handler provides tensor_pop/tensor_push function which is used inside the hook's on_save_for_backward and on_get_saved_tensor functions which are called during the context's enter function.
The synchronizer provides a sync barrier until both the offloading/reloading and the computation completes.

Reply:
@ksivaman : Can we resolve this convo? Do you have any feedback?